### PR TITLE
Refactoring

### PR DESCRIPTION
**Pull request by Tako-San**

### CODE REVIEW

**Comment by Tako-San** [comment]
Add ref-ref qualified methods

**Review by Tako-San — COMMENTED** [review]
Add ref-ref annotated methods

File: include/intersection/detail.hh [hunk 1]
```diff
@@ -277,24 +277,20 @@ bool isOverlap(Segment2D<T> &segm1, Segment2D<T> &segm2)
...
-  for (++cur; cur != end; ++cur)
-    if ((*cur) * (*prev) <= 0)
-      return false;
+  if (begin == end)
+    return true;
 
-  return true;
+  bool fst = (*begin > 0);
+  return std::none_of(std::next(begin), end,
+                      [fst](auto &&elt) { return (elt > 0) != fst || elt == 0; });
```
Comment (derzhavin3016) [inline include/intersection/detail.hh:287]: Comparing floating point using `==` operator

**Review by derzhavin3016 — CHANGES_REQUESTED** [review]

**Comment by derzhavin3016** [comment]
Remove Threshold manipulation from Vec3 class

**Comment by Tako-San** [comment]
Maybe make macros for BoundBox `min` and `max` methods?

**Comment by Tako-San** [comment]
> Maybe make macros for BoundBox `min` and `max` methods?

And add ref-ref qualified versions

**Comment by Tako-San** [comment]
Make tests build unnecessary

**Comment by Tako-San** [comment]
@derzhavin3016 finish your threshold idea and I think this pr will be ready to merge

File: include/intersection/detail.hh [hunk 2]
```diff
@@ -367,28 +354,21 @@ bool isCounterClockwise(Trian2<T> &tr)
...
-
-  for (size_t i = 1; i < 3; ++i)
-    if (auto val = dot(d, tr[i]); val < min)
-      min = val;
-    else if (val > max)
-      max = val;
-
-  return {min, max};
+  std::array<T, 3> dotArr{};
+  std::transform(tr.begin(), tr.end(), dotArr.begin(), [&d](auto &&v) { return dot(d, v); });
```
Comment (Tako-San) [inline include/intersection/detail.hh:358]: Is it possible to use std::bind_front?

File: include/intersection/detail.hh [hunk 3]
```diff
@@ -367,28 +354,21 @@ bool isCounterClockwise(Trian2<T> &tr)
...
+  return {*mmIt.first, *mmIt.second};
 }
 
 template <std::floating_point T>
 Segment3D<T> getSegment(const Triangle<T> &tr)
 {
   std::array<T, 3> lenArr{};
-  for (size_t i = 0; i < 3; ++i)
+  for (std::size_t i = 0; i < 3; ++i)
     lenArr[i] = (tr[i] - tr[i + 1]).length2();
```
Comment (Tako-San) [inline include/intersection/detail.hh:368]: Maybe use std:: transform?

Comment (Tako-San) [reply include/intersection/detail.hh:368]: Looks like it is impossible

**Comment by Tako-San** [comment]
https://github.com/106-inc/Triangles/blob/6752b0b00353d866e05ba3762e37872ca2f52329/include/primitives/vec3.hh#L155
Add && version

**Comment by Tako-San** [comment]
Looks like here works rule of zero
https://github.com/106-inc/Triangles/blob/6752b0b00353d866e05ba3762e37872ca2f52329/include/kdtree/container.hh#L24-L29

https://github.com/106-inc/Triangles/blob/6752b0b00353d866e05ba3762e37872ca2f52329/include/kdtree/container.hh#L66-L72

https://github.com/106-inc/Triangles/blob/6752b0b00353d866e05ba3762e37872ca2f52329/include/kdtree/kdtree.hh#L100-L106

**Review by Tako-San — COMMENTED** [review]
Maybe use std::bitcast instead of reinterpret_cast?

**Comment by derzhavin3016** [comment]
> Maybe use std::bitcast instead of reinterpret_cast?

@Tako-San, could you specify a place in code?

**Comment by Tako-San** [comment]
> > Maybe use std::bitcast instead of reinterpret_cast?
> 
> @Tako-San, could you specify a place in code?

Node dump

**Comment by derzhavin3016** [comment]
> Looks like here works rule of zero
> 
> https://github.com/106-inc/Triangles/blob/6752b0b00353d866e05ba3762e37872ca2f52329/include/kdtree/container.hh#L24-L29
> 
> https://github.com/106-inc/Triangles/blob/6752b0b00353d866e05ba3762e37872ca2f52329/include/kdtree/container.hh#L66-L72
> 
> https://github.com/106-inc/Triangles/blob/6752b0b00353d866e05ba3762e37872ca2f52329/include/kdtree/kdtree.hh#L100-L106
> 
> UPD: Done

In this case compiler with option `-Weffc++` signalizes that class has pointer data members but does not decalre copy ctor & assignment. So, I suppose, it's better to default them

File: include/intersection/detail.hh (hunk 1 above)
Comment (derzhavin3016) [reply include/intersection/detail.hh]: Fixed

**Comment by derzhavin3016** [comment]
> @derzhavin3016 finish your threshold idea and I think this pr will be ready to merge

Done.
Idk why, but unit test was failing on pipeline, but not on my machine. I fixed it by multiplying threshold on 100 (instead of 1000 in vector)

**Comment by Tako-San** [comment]
> > @derzhavin3016 finish your threshold idea and I think this pr will be ready to merge
> 
> Done. Idk why, but unit test was failing on pipeline, but not on my machine. I fixed it by multiplying threshold on 100 (instead of 1000 in vector)

I checked your commits, LGTM. I tried to find the reason, but I don’t know why threshold causes such troubles.

**Comment by Tako-San** [comment]
> > Looks like here works rule of zero
> > https://github.com/106-inc/Triangles/blob/6752b0b00353d866e05ba3762e37872ca2f52329/include/kdtree/container.hh#L24-L29
> > 
> > https://github.com/106-inc/Triangles/blob/6752b0b00353d866e05ba3762e37872ca2f52329/include/kdtree/container.hh#L66-L72
> > 
> > https://github.com/106-inc/Triangles/blob/6752b0b00353d866e05ba3762e37872ca2f52329/include/kdtree/kdtree.hh#L100-L106
> > 
> > UPD: Done
> 
> In this case compiler with option `-Weffc++` signalizes that class has pointer data members but does not decalre copy ctor & assignment. So, I suppose, it's better to default them

All of them is non-owning pointer, so no need to care about memory
We may return declarations if you wish

**Comment by Tako-San** [comment]
https://github.com/106-inc/Triangles/blob/052abeb0aba04620be553945bfd2f1bcee92d459/include/kdtree/container.hh#L82

Looks like we may use `= default` here

**Pull request merged**